### PR TITLE
implement fast versions of Flux

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
```julia
using DiffEqFlux, OrdinaryDiffEq, Optim, Flux, Zygote, Test

u0 = Float32[2.; 0.]
datasize = 30
tspan = (0.0f0,1.5f0)

function trueODEfunc(du,u,p,t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
t = range(tspan[1],tspan[2],length=datasize)
prob = ODEProblem(trueODEfunc,u0,tspan)
ode_data = Array(solve(prob,Tsit5(),saveat=t))

fastdudt2,p = FastChain((x,p) -> x.^3,
             FastDense(2,50,tanh),
             FastDense(50,2))
fast_n_ode = NeuralODE(fastdudt2,p,tspan,Tsit5(),saveat=t)

function fast_predict_n_ode(p)
  fast_n_ode(u0,p)
end

function fast_loss_n_ode(p)
    pred = fast_predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

dudt2 = Chain((x) -> x.^3,
             Dense(2,50,tanh),
             Dense(50,2))
n_ode = NeuralODE(dudt2,tspan,Tsit5(),saveat=t)

function predict_n_ode(p)
  n_ode(u0,p)
end

function loss_n_ode(p)
    pred = predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

_p,re = Flux.destructure(dudt2)
@test fastdudt2(ones(2),_p) ≈ dudt2(ones(2))
@test fast_loss_n_ode(p)[1] ≈ loss_n_ode(p)[1]
@test Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)[1] ≈ Zygote.gradient((p)->loss_n_ode(p)[1], p)[1]

@btime Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)
@btime Zygote.gradient((p)->fast_loss_n_ode(p)[1], p)
@btime Zygote.gradient((p)->loss_n_ode(p)[1], p)
@btime Zygote.gradient((p)->loss_n_ode(p)[1], p)
```

```
  27.272 ms (181318 allocations: 16.54 MiB)
  27.328 ms (181318 allocations: 16.54 MiB)
  262.430 ms (677868 allocations: 32.83 MiB)
  260.814 ms (677868 allocations: 32.83 MiB)
```

order of magnitude performance improvement over using Flux for neural networks

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Typical layer sizes justify the use of StaticArrays:

```julia
using StaticArrays
A = SMatrix{8,8}(randn(8,8))
b = SVector{8}(randn(8))
@btime A*b # 31.923 ns (1 allocation: 80 bytes)

A = SMatrix{16,16}(randn(16,16))
b = SVector{16}(randn(16))
@btime A*b # 46.435 ns (1 allocation: 144 bytes)

A = SMatrix{32,32}(randn(32,32))
b = SVector{32}(randn(32))
@btime A*b # 158.374 ns (1 allocation: 272 bytes)

A = randn(8,8)
b = randn(8)
@btime A*b # 97.493 ns (1 allocation: 144 bytes)

A = randn(16,16)
b = randn(16)
@btime A*b # 123.810 ns (1 allocation: 208 bytes)

A = randn(32,32)
b = randn(32)
@btime A*b # 189.783 ns (1 allocation: 336 bytes)
```

File: src/fast_layers.jl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    new{out,in,typeof(σ)}(σ),vcat(vec(initW(out, in)),initb(out))
+  end
+end
+(f::StaticDense{out,in})(x,p) where {out,in} = f.σ.(SMatrix{out,in}(uview(p,1:(out*in)))*x .+ SVector{out}(uview(p,(out*in+1):lastindex(p))))
+ZygoteRules.@adjoint function (f::StaticDense{out,in})(x,p) where {out,in}
+  W = SMatrix{out,in}(uview(p,1:(out*in)))
+  b = SVector{out}(uview(p,(out*in+1):lastindex(p)))
+  res = f.σ.(W*x .+ b)
+  function StaticDense_adjoint(ȳ)
+    W'*ȳ,vcat(vec(ȳ*res'),ȳ)
```
Comment (ChrisRackauckas) [inline src/fast_layers.jl]: @oxinabox I think this is incorrect and I was hoping to getting a double check

**Comment by ChrisRackauckas** [comment]
```julia
using DiffEqFlux, OrdinaryDiffEq, Optim, Flux, Plots

u0 = Float32[2.; 0.]
datasize = 30
tspan = (0.0f0,1.5f0)

function trueODEfunc(du,u,p,t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
t = range(tspan[1],tspan[2],length=datasize)
prob = ODEProblem(trueODEfunc,u0,tspan)
ode_data = Array(solve(prob,Tsit5(),saveat=t))

fastdudt2,p = FastChain((x,p) -> x.^3,
             FastDense(2,50,tanh),
             FastDense(50,2))
n_ode = NeuralODE(fastdudt2,p,tspan,Tsit5(),saveat=t)

function predict_n_ode(p)
  n_ode(u0,p)
end

function loss_n_ode(p)
    pred = predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

cb = function (p,l,pred) #callback function to observe training
  display(l)
  # plot current prediction against data
  pl = scatter(t,ode_data[1,:],label="data")
  scatter!(pl,t,pred[1,:],label="prediction")
  display(plot(pl))
end

# Display the ODE with the initial parameter values.
cb(p,loss_n_ode(p)...)

res1 = DiffEqFlux.sciml_train!(loss_n_ode, p, ADAM(0.01), cb = cb, maxiters = 100)
res2 = DiffEqFlux.sciml_train!(loss_n_ode, res1.minimizer, BFGS(), cb = cb)








* Status: failure (reached maximum number of iterations)

* Candidate solution
   Minimizer: [-4.39e-01, 9.18e-03, 4.12e-01,  ...]
   Minimum:   7.652895e+00

* Found with
   Algorithm:     ADAM
   Initial Point: [-1.93e-01, -1.58e-01, 2.42e-01,  ...]

* Convergence measures
   |x - x'|               = NaN ≰ 0.0e+00
   |x - x'|/|x'|          = NaN ≰ 0.0e+00
   |f(x) - f(x')|         = NaN ≰ 0.0e+00
   |f(x) - f(x')|/|f(x')| = NaN ≰ 0.0e+00
   |g(x)|                 = NaN ≰ 0.0e+00

* Work counters
   Seconds run:   10  (vs limit Inf)
   Iterations:    100
   f(x) calls:    100
   ∇f(x) calls:   100


* Status: success

* Candidate solution
   Minimizer: [-2.08e-01, 4.34e-01, 3.55e-01,  ...]
   Minimum:   4.925616e-02

* Found with
   Algorithm:     BFGS
   Initial Point: [-4.39e-01, 9.18e-03, 4.12e-01,  ...]

* Convergence measures
   |x - x'|               = 8.88e-16 ≰ 0.0e+00
   |x - x'|/|x'|          = 7.41e-16 ≰ 0.0e+00
   |f(x) - f(x')|         = 0.00e+00 ≤ 0.0e+00
   |f(x) - f(x')|/|f(x')| = 0.00e+00 ≤ 0.0e+00
   |g(x)|                 = 5.46e-01 ≰ 1.0e-08

* Work counters
   Seconds run:   27  (vs limit Inf)
   Iterations:    30
   f(x) calls:    360
   ∇f(x) calls:   360
```

This means that the whole neural ODE example is trained in 37 seconds. This is compared to:

```julia
using DiffEqFlux, OrdinaryDiffEq, Optim, Flux, Plots
u0 = Float32[2.; 0.]
datasize = 30
tspan = (0.0f0,1.5f0)

function trueODEfunc(du,u,p,t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
t = range(tspan[1],tspan[2],length=datasize)
prob = ODEProblem(trueODEfunc,u0,tspan)
ode_data = Array(solve(prob,Tsit5(),saveat=t))

dudt2 = Chain(x -> x.^3,
             Dense(2,50,tanh),
             Dense(50,2))
n_ode = NeuralODE(dudt2,tspan,Tsit5(),saveat=t)

function predict_n_ode(p)
  n_ode(u0,p)
end

function loss_n_ode(p)
    pred = predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

loss_n_ode(n_ode.p) # n_ode.p stores the initial parameters of the neural ODE

cb = function (p,l,pred) #callback function to observe training
  display(l)
  # plot current prediction against data
  #pl = scatter(t,ode_data[1,:],label="data")
  #scatter!(pl,t,pred[1,:],label="prediction")
  #display(plot(pl))
end

# Display the ODE with the initial parameter values.
cb(n_ode.p,loss_n_ode(n_ode.p)...)

res1 = DiffEqFlux.sciml_train!(loss_n_ode, n_ode.p, ADAM(0.01), cb = cb, maxiters = 100)
res2 = DiffEqFlux.sciml_train!(loss_n_ode, res1.minimizer, LBFGS(), cb = cb, maxiters = 100)
```

which takes around 5 minutes, showing that it's a real-world gain!

For reference, https://gist.github.com/ChrisRackauckas/cc6ac746e2dfd285c28e0584a2bfd320 training a neural ODE is now faster than the forward pass of torchdiffeq on Lorenz.

**Comment by ChrisRackauckas** [comment]
Fixed parameters to Float32 and now convergence is under 30 seconds in many cases:

```julia
using DiffEqFlux, OrdinaryDiffEq, Optim, Flux, Plots

using LinearAlgebra
LinearAlgebra.BLAS.set_num_threads(4)

u0 = Float32[2.; 0.]
datasize = 30
tspan = (0.0f0,1.5f0)

function trueODEfunc(du,u,p,t)
    true_A = [-0.1 2.0; -2.0 -0.1]
    du .= ((u.^3)'true_A)'
end
t = range(tspan[1],tspan[2],length=datasize)
prob = ODEProblem(trueODEfunc,u0,tspan)
ode_data = Array(solve(prob,Tsit5(),saveat=t))

fastdudt2 = FastChain((x,p) -> x.^3,
             FastDense(2,50,tanh),
             FastDense(50,2))
n_ode = NeuralODE(fastdudt2,tspan,Tsit5(),saveat=t)

function predict_n_ode(p)
  n_ode(u0,p)
end

function loss_n_ode(p)
    pred = predict_n_ode(p)
    loss = sum(abs2,ode_data .- pred)
    loss,pred
end

cb = function (p,l,pred) #callback function to observe training
  display(l)
  # plot current prediction against data
  #pl = scatter(t,ode_data[1,:],label="data")
  #scatter!(pl,t,pred[1,:],label="prediction")
  #display(plot(pl))
end

# Display the ODE with the initial parameter values.
cb(p,loss_n_ode(p)...)

res1 = DiffEqFlux.sciml_train!(loss_n_ode, p, ADAM(0.05), cb = cb, maxiters = 100)
res2 = DiffEqFlux.sciml_train!(loss_n_ode, res1.minimizer, BFGS(), cb = cb, maxiters = 100)
```

File: src/fast_layers.jl (hunk 1 above)
Comment (oxinabox) [reply src/fast_layers.jl]: best check is a test with finite differencing.

**Pull request merged**